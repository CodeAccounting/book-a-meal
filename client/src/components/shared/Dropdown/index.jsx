import React, { Component } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import LinkBtn from '../Link';

/**
 * @exports
 * @class Dropdown
 * @extends Component
 * @classdesc Creates Dropdown Component
 * @returns {JSX} Dropdown Component
 */
class Dropdown extends Component {
  /**
   * @constructor
   * @returns {JSX} Dropdown
   */
  constructor() {
    super();

    this.state = {
      showContent: false
    };
  }

  /**
   * @memberof Dropdown
   * @returns {noting} nothing
   */
  toggleDropdownContent = () => {
    this.setState({
      showContent: !this.state.showContent
    });
  }

  /**
   * @memberof Dropdown
   * @returns {JSX} Dropdown Component
   */
  render() {
    const { type, toggler, content } = this.props;
    const { showContent } = this.state;

    const dropdownClass = classNames({
      dropdown: true,
      notification: type === 'notification' || type === 'admin-notification',
      'card-dropdown': type === 'card'
    });

    const dropdownContentClass = classNames({
      show: showContent,
      'dropdown-content': true,
      'admin-dropdown-content': type === 'admin-notification',
      meal: type === 'card'
    });

    return (
      <div
        className={dropdownClass}
        onMouseEnter={this.toggleDropdownContent}
        onMouseLeave={this.toggleDropdownContent}
      >
        <LinkBtn
          clickHandler={this.toggleDropdownContent}
          id="dropdown-toggler"
          className="link"
        >
          {toggler}
        </LinkBtn>
        <div className={dropdownContentClass} id="dropdown-content">
          {content}
        </div>
      </div>
    );
  }
}

Dropdown.propTypes = {
  type: PropTypes.string.isRequired,
  toggler: PropTypes.element.isRequired,
  content: PropTypes.element.isRequired,
};

export default Dropdown;
