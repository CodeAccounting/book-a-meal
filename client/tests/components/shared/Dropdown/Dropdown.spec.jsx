import React from 'react';
import Dropdown from '../../../../src/components/shared/Dropdown';

describe('Dropdown', () => {
  it('renders correctly when type is simple notification', () => {
    const shallowWrapper = shallow(<Dropdown
      type="notification"
      toggler={<p>Click Me</p>}
      content={<div>Hey</div>}
    />);


    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('renders correctly when type is admin notification', () => {
    const shallowWrapper = shallow(<Dropdown
      type="admin-notification"
      toggler={<p>Click Me</p>}
      content={<div>Hey</div>}
    />);


    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('toggles showContent state to true on hover', () => {
    const wrapper = mount(<Dropdown
      type="notification"
      toggler={<p>Hover over Me</p>}
      content={<div>Hey</div>}
    />);

    wrapper.find('.dropdown').simulate('mouseEnter');
    expect(wrapper.state('showContent')).toEqual(true);

    wrapper.find('.dropdown').simulate('mouseLeave');
    expect(wrapper.state('showContent')).toEqual(false);

    wrapper.unmount();
  });
});
