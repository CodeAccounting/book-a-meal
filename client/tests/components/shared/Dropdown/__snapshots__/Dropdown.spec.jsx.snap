// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders correctly when type is admin notification 1`] = `
<div
  className="dropdown notification"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <Link
    className="link"
    clickHandler={null}
    dataTip={null}
    id="dropdown-toggler"
  >
    <p>
      Click Me
    </p>
  </Link>
  <div
    className="dropdown-content admin-dropdown-content"
    id="dropdown-content"
  >
    <div>
      Hey
    </div>
  </div>
</div>
`;

exports[`Dropdown renders correctly when type is simple notification 1`] = `
<div
  className="dropdown notification"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <Link
    className="link"
    clickHandler={null}
    dataTip={null}
    id="dropdown-toggler"
  >
    <p>
      Click Me
    </p>
  </Link>
  <div
    className="dropdown-content"
    id="dropdown-content"
  >
    <div>
      Hey
    </div>
  </div>
</div>
`;
